core/src/main/java/org/infinispan/persistence/async/AsyncCacheWriter.java

```diff
@@ -123,12 +116,21 @@ public Thread newThread(Runnable r) {
    @Override
    public void stop() {
       if (trace) log.tracef("Stop async store %s", this);
-      stateLock.writeLock(1);
-      state.get().stopped = true;
+      stateLock.writeLock(0);
+      stopped = true;
       stateLock.writeUnlock();
       try {
-         coordinator.join(shutdownTimeout);
-         if (coordinator.isAlive())
+         // It is safe to wait without timeout because the thread pool uses an unbounded work queue (i.e.
+         // all work handed to the pool will be accepted and eventually executed) and AsyncStoreProcessors
+         // decrement the workerThreads latch in a finally block (i.e. even if the back-end store throws
+         // a java.lang.Error). The coordinator thread can only block forever if the back-end's write() /
+         // remove() methods block, but this is no different from PassivationManager.stop() being blocked
+         // in a synchronous call to write() / remove().
+         coordinator.join();
+         // The coordinator thread waits for AsyncStoreProcessor threads to finish, so shutting down the
+         // thread pool should immediately terminate it.
+         executor.shutdown();
```

Comment:
I'm not sure the comment is correct: the coordinator waits for the threads to countdown the latch, but a thread scheduling hitch could mean the thread pool doesn't see them as finished yet. Nothing dangerous, but maybe you should replace the `isTerminated()` check with `awaitTermination(1, SECOND)` to avoid spurious errors in the log.